File: examples/quickstart.ts
/**
 * Quickstart: a full BTC ↔ USDT(EVM) swap driven entirely through the developer API.
 *
 *   npm i @hashlock-tech/sdk
 *   HASHLOCK_API_KEY=hk_test_… npx tsx examples/quickstart.ts
 *
 * This shows the API flow end-to-end. Signing is YOURS — the settlement builders return UNSIGNED
 * transactions; sign them with your wallet or HSM (see examples/fireblocks) and either broadcast via
 * client.broadcast(...) or let your custody provider broadcast. The server never holds your keys.
 */
import { HashlockClient, newSecret } from '@hashlock-tech/sdk';

const client = new HashlockClient({
  apiKey: process.env.HASHLOCK_API_KEY!,
  baseUrl: process.env.HASHLOCK_API_URL ?? 'https://api-dev.hashlock.markets/v1',
});

// Sign an unsigned settlement build with your own key/HSM, then return a txid.
// EVM → sign each tx and broadcast; TRON → sign each txID; Bitcoin → build+sign the P2WSH spend.
// See examples/fireblocks for a custody-provider reference.
declare function signAndBroadcast(build: unknown): Promise<string>;

async function main() {
  const me = await client.me();
  console.log('authenticated:', me.userId, me.scopes);

  const assets = await client.assets();
  const btc = assets.find((a) => a.chain === 'bitcoin' && a.symbol === 'BTC')!;
  const usdt = assets.find((a) => a.chain === 'ethereum' && a.symbol === 'USDT')!;

  // 1) taker: create an RFQ to sell BTC for USDT (the taker funds the long BTC leg → is the initiator).
  const rfq = await client.createRfq({
    direction: 'sell_base',
    baseAssetId: btc.id,
    baseAmount: '10000', // sats
    quoteAssetId: usdt.id,
    ttlSeconds: 3600,
  });
  console.log('rfq', rfq.id);

  // 2) maker (a different account/key) quotes it → opens a settlement thread.
  //    const { thread } = await makerClient.quoteRfq(rfq.id, '650000'); // 0.65 USDT (6 decimals)
  const threadId = '<thread id from the maker quote>';

  // 3) accept. The initiator (funds the long leg) supplies hashlock = sha256(secret).
  const { secret, hashlock } = await newSecret(); // keep `secret` private until you claim your leg
  await client.acceptTerms(threadId, { hashlock });
  // …the maker also accepts; when BOTH accept, the swap is created:
  const swap = (await client.acceptTerms(threadId)).swap!; // returns the swap once both sides accepted

  // 4) set your receive/refund addresses. Bitcoin uses the compressed pubkey (hex).
  await client.setSwapAddress(swap.id, 'ethereum', '0xYourUsdtPayoutAddress'); // taker receives USDT
  await client.setSwapAddress(swap.id, 'bitcoin', '<your btc compressed pubkey>'); // taker's BTC refund

  // 5) fund the long (BTC) leg — build → sign → broadcast.
  const fundBuild = await client.buildFund(swap.id, 'b'); // leg 'b' = BTC here
  await signAndBroadcast(fundBuild);

  // 6) once the counterparty funds the USDT leg (poll getSwap until 'counterparty_funded'),
  //    claim your USDT leg with the secret — this reveals it on-chain.
  const claimBuild = await client.buildClaim(swap.id, 'a', secret); // leg 'a' = USDT here
  await signAndBroadcast(claimBuild);

  // The counterparty then claims the BTC leg using the now-public secret. Track it live:
  for await (const s of pollSwap(client, swap.id)) {
    console.log('status:', s.status);
    if (s.status === 'counterparty_claimed' || s.status === 'refunded') break;
  }
}

async function* pollSwap(c: HashlockClient, id: string) {
  let last = '';
  for (;;) {
    const s = await c.getSwap(id);
    if (s.status !== last) { last = s.status; yield s; }
    await new Promise((r) => setTimeout(r, 4000));
  }
}

main().catch((e) => { console.error(e); process.exit(1); });

File: examples/quickstart.py
"""Quickstart: a full BTC <-> USDT(EVM) swap driven through the developer API.

    pip install hashlock-sdk
    HASHLOCK_API_KEY=hk_test_... python examples/quickstart.py

Signing is YOURS — the settlement builders return UNSIGNED transactions; sign with your wallet or HSM
(see examples/fireblocks) and either broadcast via client.broadcast(...) or let your custody provider
broadcast. The server never holds your keys.
"""
import os
import time

from hashlock import HashlockClient, new_secret


def sign_and_broadcast(build: dict) -> str:
    """Sign an unsigned settlement build with your own key/HSM and return a txid. See examples/fireblocks."""
    raise NotImplementedError("plug in your signer (wallet / Fireblocks / Copper)")


def main() -> None:
    client = HashlockClient(
        api_key=os.environ["HASHLOCK_API_KEY"],
        base_url=os.environ.get("HASHLOCK_API_URL", "https://api-dev.hashlock.markets/v1"),
    )
    me = client.me()
    print("authenticated:", me["userId"], me["scopes"])

    assets = client.assets()
    btc = next(a for a in assets if a["chain"] == "bitcoin" and a["symbol"] == "BTC")
    usdt = next(a for a in assets if a["chain"] == "ethereum" and a["symbol"] == "USDT")

    # 1) taker: sell BTC for USDT (the taker funds the long BTC leg -> is the initiator).
    rfq = client.create_rfq(
        direction="sell_base",
        base_asset_id=btc["id"],
        base_amount="10000",  # sats
        quote_asset_id=usdt["id"],
        ttl_seconds=3600,
    )
    print("rfq", rfq["id"])

    # 2) maker (a different account) quotes it -> opens a thread:
    #    thread = maker_client.quote_rfq(rfq["id"], "650000")["thread"]
    thread_id = "<thread id from the maker quote>"

    # 3) accept. The initiator supplies hashlock = sha256(secret).
    secret, hashlock = new_secret()  # keep `secret` private until you claim your leg
    client.accept_terms(thread_id, hashlock=hashlock)
    swap = client.accept_terms(thread_id)["swap"]  # created once BOTH sides have accepted

    # 4) set receive/refund addresses (Bitcoin uses the compressed pubkey hex).
    client.set_swap_address(swap["id"], "ethereum", "0xYourUsdtPayoutAddress")
    client.set_swap_address(swap["id"], "bitcoin", "<your btc compressed pubkey>")

    # 5) fund the long (BTC) leg.
    sign_and_broadcast(client.build_fund(swap["id"], "b"))

    # 6) after the counterparty funds the USDT leg, claim it with the secret (reveals it on-chain).
    sign_and_broadcast(client.build_claim(swap["id"], "a", secret))

    last = ""
    while True:
        s = client.get_swap(swap["id"])
        if s["status"] != last:
            last = s["status"]
            print("status:", last)
        if last in ("counterparty_claimed", "refunded"):
            break
        time.sleep(4)


if __name__ == "__main__":
    main()
